[P036] Using template notations within oledframedcmd,<line>,<text> that are to be reinterpreted each time before the line is displayed

uwekaditz opened the pull request:
So far within the `oledframedcmd,<line>,<text>` all template notations can be used, like system variables, or reference to a task value, but they will be interpreted only once to generate the new line content.

To use template notations that are to be **reinterpreted each time** before the line is displayed, the `<line> parameter` must be **negative**.
Additionally to the negative `<line> parameter` following conditions must also be met for the `<text> parameter`:
- use `$$<sysvar>$$` instead of `%<sysvar>%` for system variables
- use `${<TaskName>#<TaskValue>}$` instead of `[<TaskName>#<TaskValue>]` for task values

uwekaditz committed:
Merge remote-tracking branch 'upstream/mega' into UweKaditz

uwekaditz committed:
Merge branch 'UweKaditz' of https://github.com/uwekaditz/ESPEasy into UweKaditz

uwekaditz committed:
Merge remote-tracking branch 'github-desktop-TD-er/mega' into UweKaditz

uwekaditz committed:
Merge remote-tracking branch 'github-desktop-TD-er/mega' into UweKaditz

uwekaditz committed:
Merge remote-tracking branch 'github-desktop-tonhuisman/mega' into UweKaditz

uwekaditz committed:
Changes for PR5097

tonhuisman commented:
> but they will be interpreted only once to generate the new line content.

I'm trying to understand what you are saying here... 🤔

uwekaditz commented:
That means that the template notations, like system variables, or references to a task value, are interpreted (exchanged with the current value) while the oledframedcmd is transferred to the P036 plug in.
It was not possible to add or change a line content to a text that holds a template notation.
Example:
oledframedcmd,1,%systime% will always display the systime while issuing the cmd
now oledframedcmd,-1,$$systime$$ will display the updated systime while displaying the line 1

tonhuisman commented on src/_P036_FrameOLED.ino:
```diff
@@ -414,8 +417,8 @@ boolean Plugin_036(uint8_t function, struct EventStruct *event, String& string)
         {
           /* *INDENT-OFF* */
           F("SSID"),     F("SysName"),      F("IP"),   F("MAC"),         F("RSSI"),
-          F("BSSID"),    F("WiFi channel"), F("Unit"), F("SysLoad"),     F("SysHeap"),
-          F("SysStack"), F("Date"),         F("Time"), F("PageNumbers"),
+          F("BSSID"),          F("WiFi channel"),          F("Unit"),                  F("SysLoad"),                F("SysHeap"),
+          F("SysStack"),       F("Date"),                  F("Time"),                  F("PageNumbers"),
```
Hm, your formatter seems to ignore the directive to not touch the indentation...

tonhuisman commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -735,22 +745,22 @@ String P036_data_struct::create_display_header_text(eHeaderContent iHeaderConten
 {
   String newString, strHeader;
   const __FlashStringHelper *newString_f = F("%sysname%");
-  bool use_newString_f                   = true;
+  bool use_newString_f = true;
```
More formatter issues, are you using VSCode with Uncrustify, as suggested & [documented](https://espeasy.readthedocs.io/en/latest/Participate/PlatformIO.html#uncrustify) in the PlatformIO page?

uwekaditz replied:
Normally, I do.
I can try the uncrustify again.

tonhuisman replied:
Having differently configured formatting tools is rather annoying when reviewing the code-changes.

tonhuisman commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -338,32 +339,41 @@ bool P036_data_struct::plugin_write(struct EventStruct *event, const String& str
+      parseString.replace("$$", "%"); // Allow system vars to be passed in by using $$ instead of %
+      parseString.replace("${", "["); // Allow task values to be passed in by using ${ instead of [
+      parseString.replace("}$", "]"); // Allow task values to be passed in by using }$ instead of ]
+    }
```
Maybe we should put this in a function with a `byref` parameter, and also apply it to the `userdef1` and `userdef2` commands, from line 588, to make it all work the same (and update documentation, of course).

uwekaditz replied:
This would be fine, but I was afraid that a change for the header content in `userdef1` and `userdef2 `can effect running applications.
In the header just one **$** is used
`userDef1.replace('$', '%'); // Allow system vars to be passed in by using $ instead of %`

tonhuisman replied:
That could be a 4th replacement, as a fall-back, and a friendly suggestion in documentation to now use the new `$$`, `${` and `$}` token delimiters, as the old `$` will be deprecated (in the near future).

uwekaditz replied:
Last commit:
I created a function ReplaceTokenDelimiters() to be used for user defined headers and line content.
The documentation is also changed.
@tonhuisman Is there a way or extension to see the formatted output of *.repl and *.rst units in VS code?

tonhuisman replied:
> Is there a way or extension to see the formatted output of *.repl and *.rst units in VS code?

The Developers guide has a chapter on [writing documentation](https://espeasy.readthedocs.io/en/latest/Participate/PlatformIO.html#writing-documentation) that also explains how the output can be reviewed in a local browser by opening the `docs/build/html/index.html` file. After a documentation update just run the batchfile (or make html command) again from the docs folder and refresh the browser page.
That works best for me, as I don't have a 4k monitor to effectively use the preview feature of the reStructuredText plugin of VSCode. (Which implies that that is also a usable workflow 😄)

uwekaditz replied:
Thanks for this info. I've never read it before.
It really helps and works for me, too.

TD-er replied:
Just one question....
Why????
Why adding a new formatting syntax?

I have spent quite a lot of time to get rid of the many copy/paste implementations in the past where each was slightly different and thus causing unpredictable results based on where it is used.

This is in a single plugin, meaning the $ is now only used in this plugin, where in every other place in the code there are several notations to give access to variables.

tonhuisman replied:
This command not only updates the display contents, but also the internal buffer that is used to update the display after some time. Originally, the interpreted values where stored in the buffer; with this syntax the _variables_ can be stored in the buffer, to be interpreted each time the display is updated. (The part where the text is stored in the buffer is a bit hard to recognize, took me a while to find... 🤔)

IMHO this can be useful in other places, like other display plugins, too. Would have to be made a globally available function then.

TD-er replied:
OK, so we need it as an escape-character to be able to send verbatim internal variables via a command?

If so, then I can see its use, but still this then should not be part of just this plugin as it can be useful for other plugins too.
Or even to be used in commands like `sendTo`, which would trigger a command to be executed on another node via p2p.

TD-er replied:
I think an escape chafacter should be a backslash (thus `\%` for example)
The $ can then later be used for other stuff (e.g. string notations/labels or something like that)

uwekaditz replied:
That means the suggest notations would be:

- `\%<sysvar>\%` instead of `%<sysvar>%` for system variables
- `\[<TaskName>#<TaskValue>\]` instead of `[<TaskName>#<TaskValue>]` for task values

TD-er replied:
Yep, I did discuss it with Ton this morning.

So in short, this notation using escape-characters then requires a small change in the current parser and adding an extra function to strip the escape characters (e.g. when storing the values)

uwekaditz replied:
I tried the three escape char sentences.
`\%` works fine, but `\[` and `\]` did not work at all:
1. in Tools->Command I submitted `oledframedcmd,-1,'T: \[myTimer#Value\]s   '`
the logged result was
```
00:15:36.153 : (4640) Info   : HTTP: oledframedcmd,-1,'T: \[myTimer#Value\]s   '
00:15:36.164 : (4472) Info   : Enter P036_data_struct::plugin_write: oledframedcmd,-1,'T: \s   ' (27)
```
where the number in brackets is string.length()

2. in Tools->Command I submitted `oledframedcmd,-1,'T: \\[myTimer#Value\\]s   '`
the logged result was
```
00:17:49.704 : (4760) Info   : HTTP: oledframedcmd,-1,'T: \\[myTimer#Value\\]s   '
00:17:49.714 : (4592) Info   : Enter P036_data_struct::plugin_write: oledframedcmd,-1,'T: \\s   ' (28)
```
Which function is parsing inbetween submitting the command and P036_data_struct::plugin_write?

TD-er replied:
Those are not yet implemented....
The problem I would like to avoid is that we won't end up with separate formatting rules.

The replacement of system variables etc. are done in the `String parseTemplate(String& tmpString)` function.

tonhuisman replied:
So that would end up in a couple of tasks:
- Don't process variables wrapped in escape sequences when processing template variables (replace by placeholders before processing, restore after processing)
- Provide a `fixupEscapedVariables()` (better name to be proposed) function that removes the escape characters
- Call that function from plugins that need the escaped variables turned into normal variables for further processing (mostly display related, I think, though there may be other uses)

TD-er replied:
`stripEscapeCharacters` would probably be a better name for the function.

Other use case to use escape characters is when sending something via p2p to another node where the other node should fetch its local system variables

uwekaditz replied:
Just now I comitted a version with the new functions `stripEscapeCharacters()` and `hasEscapeCharacters()`

TD-er replied:
What happens when you try to send some string with those escape characters to log?
For example:
```
logentry,"Systime: \%systime\% without escapechar: %systime%"
```

uwekaditz committed:
Code uncrustified

uwekaditz committed:
Function ReplaceTokenDelimiters() to be used for user defined headers and line content

tonhuisman commented on docs/source/Plugin/P036_commands.repl:
```diff
@@ -85,15 +85,17 @@
-        Use ``$<sysvar>$`` instead of ``%<sysvar>%`` to use system variables.
+        * use ``$$<sysvar>$$`` instead of ``%<sysvar>%`` for system variables
+        * use ``${<TaskName>#<TaskValue>}$`` instead of ``[<TaskName>#<TaskValue>]`` for task values
+        **Note:** The use of ``$<sysvar>$`` instead of ``%<sysvar>%`` to use system variables will be depricated in near future.
```
Small typo: (2x) `depricated in near future.` -> `deprecated in the near future.`

uwekaditz replied:
Done, see last commit

tonhuisman commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -284,6 +284,23 @@ bool P036_data_struct::init(taskIndex_t      taskIndex,
   return isInitialized();
 }
 
+String P036_data_struct::ReplaceTokenDelimiters(const String str, const bool ForHeaderOnly = false)
```
To make it more lean, you could change `const String str` to `String &str`, so no strings are exchanged via the stack (and an extra copy created), but only a pointer passed, as the intent of these replacements is to be as close to 'in-place' as possible.

uwekaditz replied:
Done, see last commit

tonhuisman commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -284,6 +284,23 @@ bool P036_data_struct::init(taskIndex_t      taskIndex,
 
+String P036_data_struct::ReplaceTokenDelimiters(const String str, const bool ForHeaderOnly = false)
+{
+  String s = str;
```
This copy is not needed.

uwekaditz replied:
Done, see last commit

tonhuisman commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -590,15 +605,15 @@ bool P036_data_struct::plugin_write(struct EventStruct *event, const String& str
 
       case p036_subcommands_e::userdef1:
       {
-        userDef1 = parseStringKeepCase(string, 3);
-        userDef1.replace('$', '%'); // Allow system vars to be passed in by using $ instead of %
+        userDef1 = ReplaceTokenDelimiters(parseStringKeepCase(string, 3), true); // Allow system vars and task values to be passed to the
+                                                                                 // user defined header1
         break;
       }
 
       case p036_subcommands_e::userdef2:
       {
-        userDef2 = parseStringKeepCase(string, 3);
-        userDef2.replace('$', '%'); // Allow system vars to be passed in by using $ instead of %
+        userDef2 = ReplaceTokenDelimiters(parseStringKeepCase(string, 3), true); // Allow system vars and task values to be passed to the
```
the result of `parseString...` should be assigned to the `userDefN` var, and then that should get the replacements.

uwekaditz replied:
Done, see last commit

tonhuisman commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -800,8 +815,6 @@ String P036_data_struct::create_display_header_text(eHeaderContent iHeaderConten
-      strHeader  = F("page ");
-      strHeader += (currentFrameToDisplay + 1);
       strHeader       = F("page ");
       strHeader      += (currentFrameToDisplay + 1);
```
This can be optimized to:
```c++
strHeader       = strformat(F("page %d"), currentFrameToDisplay + 1);
```

uwekaditz replied:
Done, see last commit

uwekaditz committed:
Changes requested by tonhuisman

uwekaditz committed:
Documentation formatted

uwekaditz committed:
NEW: functions stripEscapeCharacters() and hasEscapeCharacters()

In parseTemplate()
parseSystemVariables is not executed if EscapeCharacters \% is in tmpString,
findNextDevValNameInString() is not executed if EscapeCharacters \[ or \] are in tmpString

tonhuisman commented on src/src/Helpers/StringParser.cpp:
```diff
@@ -56,12 +70,13 @@ String parseTemplate_padded(String& tmpString, uint8_t minimal_lineSize, bool us
   }
-  parseSystemVariables(tmpString, useURLencode);
-
+  if (!hasEscapeCharacters(&tmpString, "\\%"))
+    parseSystemVariables(tmpString, useURLencode);
 
   int startpos = 0;
   int lastStartpos = 0;
   int endpos = 0;
+  if (!hasEscapeCharacters(&tmpString, "\\[") && (!hasEscapeCharacters(&tmpString, "\\]")))
```
So if there are escape characters present, the entire template won't be parsed?
That's not how it should go, IMHO.
- `parseSystemVariables()` and `parseStandardConversions()` should each take care of the `\%` escape with a separate alternative to avoid conflicts by unexplicit processing-order:
- `parseTemplate_padded()` should handle `\[` and `\]`:
- `parse_string_commands()` should handle the escapes `\{`, `\}` and `\(`, `\)`:
- Procedure:
  - First replace each escape to be excluded by f.e. an ASCII character below 0x20, like 0x04 and 0x05 (AFAIK, 0x01 .. 0x03 are being used ~somewhere~ in parse_string_commands() already, ~have to check~), take care not to have conflicting (duplicate) replacements!.
  - The process then should replace all (remaining) variables as usual.
  - Before returning the result, the initial escape replacements should be reverted.

(Still checking if there are more functions to adjust)

NB: You really should be using the `F()` macro on literal strings.

tonhuisman commented on src/src/Helpers/StringParser.cpp:
```diff
@@ -26,6 +26,20 @@
+{
+  str->replace("\\%", "%"); // Allow system vars to be passed in by using \% instead of %
+  str->replace("\\[", "["); // Allow task values to be passed in by using \[ instead of [
+  str->replace("\\]", "]"); // Allow task values to be passed in by using \] instead of ]
```
Additionally, the escapes for `\{`, `\}` and `\(` and `\)` should be handled here (see previous remark)

tonhuisman commented on src/src/Helpers/StringParser.cpp:
```diff
@@ -26,6 +26,20 @@
  \*********************************************************************************************/
+// Checks str for the given escaped characters
+// So far \\% \\[ and \\] are used (all with single backslash!)
+bool hasEscapeCharacters(String *str, const String EscapeChar)
+{
+return (str->indexOf(EscapeChar)>=0);
+}
```
That `const String` can be a `const __FlashStringHelper *`, that will require those arguments to have the `F()` macro applied.

And the `return` can be indented (please don't Uncrustify this file, that'll cause hundreds of spacing changes...)

tonhuisman commented on src/src/Helpers/StringParser.cpp:
```diff
@@ -26,6 +26,20 @@
  \*********************************************************************************************/
+// Checks str for the given escaped characters
+// So far \\% \\[ and \\] are used (all with single backslash!)
+bool hasEscapeCharacters(String *str, const String EscapeChar)
```
Another, little more generic, remark on this function definition.
You are using explicit pointer definition(s) for the String at hand, and though that's not _wrong_, using a reference (`&`) is more efficient, especially in an embedded environment. References are fairly often used in ESPEasy, also in favor of creating a new pointer that gets passed via the stack, and removed afterward, instead of only the reference(-pointer).
I've seen this in other parts of your code too, and I don't want to start a discussion on it, it's more a thing I noticed, and wondered about. I assume it's the C-programming background you have been using for (quite) some time, but as said, for embedded programming, references are a somewhat better choice. I do appreciate the more explicit use of the `&` pointer-notation when passing the argument, but efficiency gets a higher score for me 😅

TD-er replied:
One good argument to use references (`&`) instead of pointers is that you put the responsibility to the caller to make sure it a valid constructed object and no need to check for null pointers.
Less checks mean smaller code and it is more predictable.

uwekaditz replied:
> Another, little more generic, remark on this function definition. You are using explicit pointer definition(s) for the String at hand, and though that's not _wrong_, using a reference (`&`) is more efficient, especially in an embedded environment. References are fairly often used in ESPEasy, also in favor of creating a new pointer that gets passed via the stack, and removed afterward, instead of only the reference(-pointer). I've seen this in other parts of your code too, and I don't want to start a discussion on it, it's more a thing I noticed, and wondered about. I assume it's the C-programming background you have been using for (quite) some time, but as said, for embedded programming, references are a somewhat better choice. I do appreciate the more explicit use of the `&` pointer-notation when passing the argument, but efficiency gets a higher score for me 😅

Thanks for the detailed information about `byref` parameters.
I'll keep it in mind for further PRs.

uwekaditz committed:
More advanced parsing as requested by @tonhuisman

parse_string_commands(), parseSystemVariables() and parseTemplate_padded() do handle escaped characters internally
Additionally, the escapes for \{, \} and \( and \) are handled
F() macro used
new functions using a byref parameter

uwekaditz commented:
Why do I always hit the program size limitation? :scream:
```
collection_D_ESP8266_4M1M_VCC
Error: The program size (1044852 bytes) is greater than maximum allowed (1044464 bytes)
```
Should I disable the new parsing functions with `P036_LIMIT_BUILD_SIZE`?

tonhuisman commented:
> Why do I always hit the program size limitation?

You keep adding code to P036 😆, and that's used in nearly all builds, but some builds, like Collection C and D, are really close to the limit for ESP8266, that's why we disable a lot of new features for ESP8266 builds: it's too much work to try and make them fit again, especially when so much time _has_ already been spent on that. And ESP8266 is an end-of-life/frozen platform in Espressif' plans...

For Collection D USES_P094 should only be enabled for ESP32, not sure if that's already merged (wrap with `#ifdef ESP32 / #endif` in `define_plugin_sets.h` if not)

tonhuisman commented on src/src/Helpers/StringParser.cpp:
```diff
@@ -70,13 +74,24 @@ String parseTemplate_padded(String& tmpString, uint8_t minimal_lineSize, bool us
+    MaskEscapedBracket = static_cast<char>(0x11); // ASCII 0x11 = Device control 1
+    tmpString.replace(F("\\["), MaskEscapedBracket);
+    MaskEscapedBracket = static_cast<char>(0x12); // ASCII 0x12 = Device control 2
+    tmpString.replace(F("\\]"), MaskEscapedBracket);
```
You're reusing 0x11/0x12 also for the `\(` and `\)` replacements, but that should better be avoided, as the call-order of template processing functions is not guaranteed, per sé, and _might_ change in the future.

uwekaditz commented:
> You keep adding code to P036

Yes, this this the favoured plug in for my applications. While using it there are sometimes new ideas how to improve it.

uwekaditz committed:
Reuse of 0x11/0x12 also for the \( and \)  changed to 0x05/0x06

For CollectionD USES_P094 should only be enabled for ESP32 according to @tonhuisman

uwekaditz committed:
Merge remote-tracking branch 'github-desktop-tonhuisman/mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Documentation updated

tonhuisman commented on docs/source/Plugin/P036_commands.repl:
```diff
@@ -42,14 +42,13 @@
-        * use ``${<TaskName>#<TaskValue>}$`` instead of ``[<TaskName>#<TaskValue>]`` for task values
+        Additionally to the **negative** <line> parameter following conditions must also be met:
+                * use ``\%<sysvar>\%`` instead of ``%<sysvar>%`` for system variables
+                * use ``\[<TaskName>#<TaskValue>\]`` instead of ``[<TaskName>#<TaskValue>]`` for task values
```
You could also add an example for `\{some_function\}` to make the example exhaustive 😸

uwekaditz replied:
Done!

uwekaditz committed:
Documentation updated for functions

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

tonhuisman commented on src/src/Helpers/StringConverter.cpp:
```diff
@@ -1317,9 +1317,21 @@ void parseSingleControllerVariable(String            & s,
 
 void parseSystemVariables(String& s, bool useURLencode)
 {
+  String MaskEscapedPercent = "";
```
Strings are created with empty content by default, no need to initialize them with the (undesired) plain C string "" value.

uwekaditz replied:
Done!

tonhuisman commented on src/src/Helpers/StringParser.cpp:
```diff
@@ -26,6 +26,24 @@
  \*********************************************************************************************/
+bool hasEscapedCharacter(String& str, const char EscapeChar)
+{
+  String EscStr = F("\\");
+  EscStr += EscapeChar;
```
This could be combined into a single line like this, and also save a few bytes in size:
`String EscStr = concat(F("\\"), EscapeChar);`

uwekaditz replied:
Done!

tonhuisman commented on src/src/Helpers/StringParser.cpp:
```diff
@@ -26,6 +26,24 @@
+
+void stripEscapeCharacters(String& str)
+{
+  str.replace(F("\\%"), "%"); // Allow system vars to be passed in by using \% instead of %
+  str.replace(F("\\["), "["); // Allow task values to be passed in by using \[ instead of [
+  str.replace(F("\\]"), "]"); // Allow task values to be passed in by using \] instead of ]
+  str.replace(F("\\{"), "{"); // Allow commands to be passed in by using \{ instead of {
+  str.replace(F("\\}"), "}"); // Allow commands to be passed in by using \} instead of }
+  str.replace(F("\\("), "("); // Allow commands to be passed in by using \( instead of (
+  str.replace(F("\\)"), ")"); // Allow commands to be passed in by using \) instead of )
```
The replacement strings should also be using the `F()` macro.

Using an array of chars and a `for` loop is even more compact, something like this:
``` c++
  const uint8_t braces[]     = { '%', '[', ']', '{', '}', '(', ')' };
  constexpr uint8_t nrbraces = NR_ELEMENTS(braces);

  for (uint8_t i = 0; i < nrbraces; ++i) {
    str.replace(strformat(F("\\%c"), braces[i]), strformat(F("%c"), braces[i]));
  }
```
On ESP32 saves ca. 50 bytes, on ESP8266 saves ca. 40 bytes in bin size.

TD-er replied:
Why not simply do this:
```c++
    str.replace(strformat(F("\\%c"), braces[i]), String(braces[i]));
```

or:
```c++
    str.replace(concat(F("\\"), braces[i]), String(braces[i]));
```

or:
```c++
    const String s(concat(F("\\"), braces[i]));
    str.replace(s, s.substring(2));
```

uwekaditz replied:
Done and bugs fixed:
`const char braces[]     = { '%', '[', ']', '{', '}', '(', ')' };`
and
`str.replace(s, s.substring(1));`

tonhuisman commented on src/src/Helpers/StringParser.cpp:
```diff
@@ -58,10 +76,22 @@ String parseTemplate_padded(String& tmpString, uint8_t minimal_lineSize, bool us
   int lastStartpos = 0;
   int endpos = 0;
+  bool mustReplaceEscapedSquareBracket = false;
+  String MaskEscapedBracket = "";
```
Another unneeded string init.

uwekaditz replied:
Done!

tonhuisman commented:
A few more comments, as we have the intent to soon merge this, but a few small improvements can be made 👼

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Removed unneeded string init, code simplified

uwekaditz committed:
BUG: simplified stripEscapeCharacters() did not work

uwekaditz commented:
The check for `normal_IRext_no_rx_ESP8266_4M2M` failed due to program size (304 bytes too big!).
Can I disable the new EscapedCharacter feature by the compiler switch `PLUGIN_BUILD_IR_EXTENDED_NO_RX` or `LIMIT_BUILD_SIZE` or should I create a new compiler switch to disable the new EscapedCharacter feature?

tonhuisman commented:
The build failure is most likely caused by the increased size of the HeatpumpIR library, that's been updated by the last merge TD-er did.
As an alternative you could disable a plugin from the STABLE set, specifically for this build set, f.e. P040 RFID - ID12LA/RDM6300, as that's quite unlikely to be used combined with the IR features in this build.
Around line 2081 in `define_plugin_sets.h`:
```
  #ifdef USES_P040
    #undef USES_P040  // RFID - ID12LA/RDM6300
  #endif
```
This brings down the build size on Windows to 99.9%, ~600 bytes below the max.

uwekaditz committed:
Disable P040 (RFID - ID12LA/RDM6300) for PLUGIN_BUILD_IR_EXTENDED_NO_RX  due to limited size

TD-er commented:
Hmm Ton I don't think I can find anything left to nitpick about ;)

I guess we should merge it, right?

tonhuisman commented:
> I guess we should merge it, right?

Yes, lgtm
